wasm-text.md

```diff
@@ -275,17 +284,47 @@ Other desugarings are either left for runtime or expressed as macros (for now).
 ### Unfolding Abbreviations
 
 ```k
-    syntax Stmts ::= unfoldStmts  ( Stmts )       [function]
-    syntax Defns ::= unfoldDefns  ( Defns )       [function]
-                   | #unfoldDefns ( Defns , Int ) [function]
- // --------------------------------------------------------
+    syntax Stmts ::= unfoldStmts  ( Stmts )            [function]
+    syntax Defns ::= unfoldDefns  ( Defns )            [function]
+                   | #unfoldDefns ( Defns , Int, Map ) [function]
+ // -------------------------------------------------------------
     rule unfoldStmts(( module OID:OptionalId DS) SS) => ( module OID unfoldDefns(DS) ) unfoldStmts(SS)
     rule unfoldStmts(.Stmts) => .Stmts
     rule unfoldStmts(S SS) => S unfoldStmts(SS) [owise]
 
-    rule unfoldDefns(DS) => #unfoldDefns(DS, 0)
-    rule #unfoldDefns(.Defns, _) => .Defns
-    rule #unfoldDefns(D:Defn DS, I) => D #unfoldDefns(DS, I) [owise]
+    rule unfoldDefns(DS) => #unfoldDefns(DS, 0, .Map)
+    rule #unfoldDefns(.Defns, _, _) => .Defns
+    rule #unfoldDefns(D:Defn DS, I, M) => D #unfoldDefns(DS, I, M) [owise]
+```
+
+#### Types
+
+A type can be be declared implicitly when a `TypeUse` is a `TypeDecls`.
+In this case it will allocate a type when the type is not in the current module instance.
+
+```k
+    rule #unfoldDefns((type ID:Identifier (func TDECLS)) DS, I, M)
+      => (type ID (func TDECLS)) #unfoldDefns(DS, I, M [ asFuncType(TDECLS) <- ID])
+      requires notBool asFuncType(TDECLS) in_keys(M)
+
+    rule #unfoldDefns((type               (func TDECLS)) DS, I, M)
+      => (type #freshId(I) (func TDECLS)) #unfoldDefns(DS, I +Int 1, M [ asFuncType(TDECLS) <- #freshId(I)])
+      requires notBool asFuncType(TDECLS) in_keys(M)
+
+    rule #unfoldDefns((type OID (func TDECLS)) DS, I, M)
+      => (type OID (func TDECLS)) #unfoldDefns(DS, I, M)
+      requires         asFuncType(TDECLS) in_keys(M)
+
+    rule #unfoldDefns(( func OID:OptionalId TDECLS:TypeDecls LOCALS:LocalDecls BODY:Instrs ) DS, I, M)
+      => (func OID (type {M [asFuncType(TDECLS)]}:>Identifier) TDECLS LOCALS BODY)
+         #unfoldDefns(DS, I, M)
+      requires         asFuncType(TDECLS) in_keys(M)
+
+    rule #unfoldDefns(( func OID:OptionalId TDECLS:TypeDecls LOCALS:LocalDecls BODY:Instrs ) DS, I, M)
+      => (type #freshId(I) (func TDECLS))
+         (func OID (type #freshId(I)) TDECLS LOCALS BODY)
+         #unfoldDefns(DS, I +Int 1, M [asFuncType(TDECLS) <- #freshId(I)])
+      requires notBool asFuncType(TDECLS) in_keys(M)
```

Comment:
Why do we use `#freshId` here instead of just taking `I`? I thought the purpose of `I` is that it would be a fresh identifier not scene anywhere yet?

Reply:
It's an `Int`, not an `Identifier`. Using an `Int` to type the function is allowed (selecting the `X`th type of the module as the type for the function). But the actual indices we give the things in a module are strict. The first `type` to appear must get index 0, the next one index 1, etc. The production for type declarations is `(type OID (func TDECLS)`. It can take an identifier, but we can't assign it an index.

We could skip using identifiers, but then we would have to keep track of exactly which index the type will get, but that would require some extra bookkeeping. The same type can be declared more than once (but can't be generated implicitly more than once), so we can't just look at the size of the map.

Reply:
In the end I did add that extra bookkeeping, see c1d0558